### PR TITLE
Module3

### PR DESCRIPTION
**Pull request by IharBakhanovich**
The Module3 to review.

### CODE REVIEW

File: certificateshop-logic/pom.xml [hunk 1]
```diff
@@ -13,18 +13,43 @@
...
+<!--        </dependency>-->
         <dependency>
-            <groupId>com.zaxxer</groupId>
-            <artifactId>HikariCP</artifactId>
+            <groupId>org.springframework.data</groupId>
+            <artifactId>spring-data-bom</artifactId>
+            <version>${spring-data-bom.version}</version>
+            <scope>import</scope>
+            <type>pom</type>
+        </dependency>
```
Comment (IvanYarochkin) [inline certificateshop-logic/pom.xml]: You can't use spring data in a scope of this module, why did you add this dependency?

File: certificateshop-logic/pom.xml [hunk 2]
```diff
@@ -39,17 +64,52 @@
...
         <dependency>
             <groupId>com.fasterxml.jackson.core</groupId>
             <artifactId>jackson-annotations</artifactId>
-            <version>2.12.4</version>
+            <version>${jackson-annotations.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.hateoas</groupId>
+            <artifactId>spring-hateoas</artifactId>
+            <version>0.23.0.RELEASE</version>
```
Comment (IvanYarochkin) [inline certificateshop-logic/pom.xml]: Move version to properties

File: certificateshop-logic/src/main/java/com/epam/esm/configuration/LogicConfig.java [hunk 3]
```diff
@@ -23,6 +27,12 @@
  * The LogicConfig class.
  */
 @Configuration
+@EnableJpaRepositories(basePackages = {
+        "com.epam.esm.repository"
+})
+@EntityScan(basePackages = {
+        "com.epam.esm.model"
+})
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/configuration/LogicConfig.java]: You have only one package, you can use `@EntityScan("com.epam.esm.model")`. Check if it's good for EnableJpaRepositories annotation as well

File: certificateshop-logic/src/main/java/com/epam/esm/dao/ListToSetConverter.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.epam.esm.dao;
+
+import java.sql.ResultSet;
+import java.util.List;
+
+public interface ListToSetConverter {
+    ResultSet getResultSet(List<String> headers, List<List<Object>> data) throws Exception;
+}
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/ListToSetConverter.java]: This is so weird method name for ListToSetConverter because I expect to see some convert... method in the convertor rather than getResultSet

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Stores names of the database tables columns, which are used by {@link GiftCertificateExtractor}
+ * {@link CertificateTagMapper} and {@link GiftCertificateMapper} to fetch data from a ResultSet.
+ */
+public class ColumnNames {
+    private ColumnNames() {
+    }
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java]: Move constructor after the below class variables.

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Stores names of the database tables columns, which are used by {@link GiftCertificateExtractor}
+ * {@link CertificateTagMapper} and {@link GiftCertificateMapper} to fetch data from a ResultSet.
+ */
+public class ColumnNames {
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java]: Util classes which have only static methods should be final as well because there are no methods to override

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+     * @return
+     */
+    public static<T> Map<String, String> createNextParameters(List<T> entities, int pageNumber, int amountOfEntitesOnThePage) {
+        int nextPageNumber = 0;
+        Map<String, String> paramsNext = new HashMap<>();
+        if (amountOfEntitesOnThePage == entities.size()) {
+            nextPageNumber = pageNumber + 1;
+        } else {
+            nextPageNumber = pageNumber;
+        }
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java:65]: Could be replaced to `nextPageNumber  = amountOfEntitesOnThePage == entities.size() ? pageNumber + 1 : pageNumber;`

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    public static<T> Map<String, String> createPrevParameters(List<T> certificates, int pageNumber, int amountOfEntitiesOnThePage) {
+        Map<String, String> paramsPrev = new HashMap<>();
+        int prevPageNumber = 0;
+        if (pageNumber > 0) {
+            prevPageNumber = pageNumber - 1;
+        } else {
+            prevPageNumber = pageNumber;
+        }
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java:79]: Could be replaced to int prevPageNumber = pageNumber > 0 ? prevPageNumber - 1 : prevPageNumber;

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        paramsPrev.put(ColumnNames.PAGE_NUMBER_PARAM_NAME, String.valueOf(prevPageNumber));
+        setLimit(amountOfEntitiesOnThePage, paramsPrev);
+        return paramsPrev;
+    }
+
+    private static void setLimit(int amountOfEntitiesOnThePage, Map<String, String> paramsNext) {
+        if (amountOfEntitiesOnThePage != 0) {
+            paramsNext.put(ColumnNames.AMOUNT_OF_ENTITIES_ON_THE_PAGE_PARAM_NAME,
+                    String.valueOf(amountOfEntitiesOnThePage));
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java]: Don't move it to the next line

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    /**
+     * Sets 'offset' and 'limit' query parameters if they are not set.
+     *
+     * @param parameters the map to validate.
+     * @return parameters of the query.
+     */
+    public static Map<String, String> validateParameters(Map<String, String> parameters,
+                                                         String defaultAmountEntitiesOnThePage) {
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java]: Don't move it to the next line

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/OrderExtractor.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+                    if (resultSet.getString(ColumnNames.TABLE_USERORDER_CERTIFICATE_COLUMN_CERTIFICATEINJSON) != null) {
+                        GiftCertificate giftCertificate = certificateInJsonMapper.mapRow(resultSet, resultSet.getRow());
+                        orders.stream()
+                                .filter(order -> order.getId() == currentId)
+                                .findAny().ifPresent(order -> order.getCertificates().add(giftCertificate));
+                    }
+                }
+            }
+        }
+        return orders;
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/OrderExtractor.java]: Divide this method into many private methods because there are sooo many if statements!

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/UserExtractor.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+                    Order newOrder = createNewOrderFromResultSetLine(resultSet);
+                    user.getOrders().add(newOrder);
+                    users.add(user);
+                    addGiftCertificateToNewOrderOfUsersOrders(resultSet, users, currentId, currentOrderId, newOrder);
+                } else {
+                    users.add(user);
+                }
+            }
+        }
+        return users;
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/UserExtractor.java]: Divide this method into many private methods because there are sooo many if statements!

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/UserExtractor.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+                resultSet.getTimestamp(ColumnNames.TABLE_USERORDER_COLUMN_CREATE_DATE).toLocalDateTime(),
+                resultSet.getString(ColumnNames.TABLE_USERORDER_COLUMN_NAME),
+                new ArrayList<>()
+        );
+    }
+
+    private void addGiftCertificateToCorrespondingOrderOfUserOrders(ResultSet resultSet, List<User> users, long currentId, long currentOrderId) throws SQLException {
+        if (resultSet.getString(ColumnNames.TABLE_USERORDER_CERTIFICATE_COLUMN_CERTIFICATEINJSON) != null) {
+            GiftCertificate giftCertificate = certificateInJsonMapper.mapRow(resultSet, resultSet.getRow());
+            users.stream().filter(user2 -> user2.getId() == currentId)
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/UserExtractor.java]: Why do you use exactly `user2` name? Where is the first user?

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+    private List<GiftCertificate> getEntities(List<List<Object>> result) {
+        ResultSet resultSet;
+        List<GiftCertificate> certificates = new ArrayList<>();
+        try {
+            resultSet = listToSetConverter.getResultSet(CERTIFICATE_HEADERS, result);
+            certificates = giftCertificateExtractor.extractData(resultSet);
+        } catch (Exception exception) {
+            throw new RuntimeException(exception);
+        }
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java]: When can the mentioned exception be thrown?

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+     * @param name is the {@link String} to find.
+     * @return {@link Optional<GiftCertificate>}.
+     */
+    @Override
+    public Optional<GiftCertificate> findByName(String name) {
+        return entityManager
+                .createQuery("select c from Certificate c where c.name = :name", GiftCertificate.class)
+                .setParameter("name", name)
+                .getResultList().stream().findFirst();
+//        return giftCertificateRepository.findByName(name);
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java]: Remove all unnecessary comments!

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    /**
+     * Finds certificate without {@link CertificateTag} by its name.
+     *
+     * @param name the name to find by.
+     * @return {@link Optional<GiftCertificate>}.
+     */
+    @Override
+    public Optional<GiftCertificate> findCertificateWithoutTagsByName(String name) {
+        return Optional.empty();
+    }
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java]: Why do you keep such empty methods? What do you know about SOLID (exactly **I** letter)?

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+    /**
+     * Removes the {@link Order} object from the database.
+     *
+     * @param id is the value of the {@link long} to find.
+     */
+    @Override
+    public void delete(long id) {
+
+    }
+
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java]: Why do you keep such empty methods? What do you know about SOLID (exactly I letter)?

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+    private List<Order> getEntities(List<List<Object>> result) {
+        ResultSet resultSet;
+        List<Order> orders = new ArrayList<>();
+        try {
+            resultSet = listToSetConverter.getResultSet(ORDER_HEADERS, result);
+            orders = orderExtractor.extractData(resultSet);
+        } catch (Exception exception) {
+            throw new RuntimeException(exception);
+        }
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java]: Why do you use default exceptions instead of your own?

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+    @Override
+    public Optional<Order> findByName(String name) {
+        Query query = entityManager.createNativeQuery(FIND_ENTITY_BY_NAME_SQL).setParameter(1, name);
+        List<Object[]> resultList = query.getResultList();
+        List<List<Object>> result = convertListOfArrayToListOfLists(resultList);
+        List<Order> users = getEntities(result);
+        if (users.isEmpty()) {
+            return Optional.empty();
+        }
+        return users.stream().findFirst();
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java:170]: You can replace to `return users.isEmpty() ? Optional.empty() : users.stream().findFirst();`

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    private List<List<Object>> convertListOfArrayToListOfLists(List<Object[]> resultList) {
+        List<List<Object>> result = new ArrayList();
+        for (Object[] objects : resultList) {
+            List<Object> listOfObjects = new ArrayList<>(Arrays.asList(objects));
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java:176]: It's better to name this variable as `objectList`

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaUserDaoImpl.java [hunk 21]
```diff
@@ -0,0 +1,186 @@
+package com.epam.esm.dao.impl.jpa;
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaUserDaoImpl.java:1]: Why do you use native query in almost all repository methods? There are more easier approaches to update or create entities.

File: certificateshop-logic/src/main/java/com/epam/esm/model/converter/CertificateFromJsonConverter.java [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+        Gson gson = new Gson();
+        return gson.toJson(giftCertificate);
+    }
+
+    @Override
+    public GiftCertificate convertToEntityAttribute(String s) {
+        Gson gson = new Gson();
+        return gson.fromJson(s, GiftCertificate.class);
+    }
+}
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/model/converter/CertificateFromJsonConverter.java]: What does `s` mean?

File: certificateshop-logic/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java [hunk 23]
```diff
@@ -68,21 +71,41 @@ public GiftCertificate createCertificate(GiftCertificate giftCertificate) {
...
+                        certificateTags1.add(certificateTag1.get());
+                    }
+                }
+            }
+        } else {
+            if (idCertificateFromDB!=0) {
+                certificateTags1 = certificateDAO.findById(idCertificateFromDB).get().getTags();
             }
         }
+        giftCertificate.setTags(certificateTags1);
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java]: Divide into small private methods because there are many if statements

File: certificateshop-logic/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
+    private final OrderDao orderDao;
+    private final CertificateValidator certificateValidator;
+    private final TagValidator tagValidator;
+    private final UserValidator userValidator;
+    private final Translator translator;
+    private final OrderValidator orderValidator;
+
+    @Autowired
+
+    public OrderServiceImpl(UserDao userDao, CertificateDao certificateDAO, TagDao tagDAO, OrderDao orderDao,
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java]: remove space

File: certificateshop-logic/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,260 @@
...
+                        translator.toLocale("SOME_ID_SHOULD_NOT_BE_LESS_THAN_ONE"), "certificateId"));
+            } else if (!certificateDAO.findById(certificate.getId()).isPresent()) {
+                errorMessage.add(String.format(translator.toLocale(
+                        "THERE_IS_NO_A_CERTIFICATE_WITH_SUCH_AN_ID_IN_DATABASE"), certificate.getId()));
+            }
+        }
+        if (!errorMessage.isEmpty()) {
+            throw new EntityNotFoundException(ERROR_CODE_ENTITY_NOT_FOUND + ERROR_CODE_ORDER_NOT_VALID, errorMessage);
+        }
+    }
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java]: Divide into small private methods, because of IFs

File: certificateshop-logic/src/main/java/com/epam/esm/service/impl/UserServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,269 @@
...
+                User createdUser = createNewUser(user);
+                for (Order order : user.getOrders()) {
+                    addOrdersToUser(user, createdUser, order, errorMessage);
+                }
+            }
+        } else {
+            errorMessage.add(translator.toLocale("THE_FIELD_NICKNAME_SHOULD_BE_NOT_EMPTY"));
+            throw new MethodArgumentNotValidException(ERROR_CODE_DUPLICATE + ERROR_CODE_USER_NOT_VALID, errorMessage);
+        }
+        return userDao.findByName(user.getNickName()).get();
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/service/impl/UserServiceImpl.java]: Do you have a test for this method? If so, do you created tests for ALL cases?

Comment (IvanYarochkin) [reply certificateshop-logic/src/main/java/com/epam/esm/service/impl/UserServiceImpl.java]: Divide this method into many small private methods because of IFs

File: certificateshop-logic/src/main/java/com/epam/esm/service/impl/UserServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,269 @@
...
+        checkId(userId);
+        Optional<User> user = userDao.findById(userId);
+        if (!user.isPresent()) {
+            List<String> errorMessage = new ArrayList<>();
+            errorMessage.add(String.format(translator
+                    .toLocale("THERE_IS_NO_A_USER_WITH_SUCH_AN_ID_IN_DATABASE"), userId));
+            throw new EntityNotFoundException(ERROR_CODE_ENTITY_NOT_FOUND + ERROR_CODE_USER_NOT_VALID, errorMessage);
+        } else {
+            return user.get();
+        }
```
Comment (IvanYarochkin) [inline certificateshop-logic/src/main/java/com/epam/esm/service/impl/UserServiceImpl.java]: Just `return user.get();` without else

**Review by IvanYarochkin — CHANGES_REQUESTED** [review]
I haven't found new tests

File: certificateshop-logic/pom.xml (hunk 1 above)
Comment (IharBakhanovich) [reply certificateshop-logic/pom.xml]: there are some methods which out of the functionlal. Those methods use spring data. I studied spring data too. That is the point whz i use this dependency.

File: certificateshop-logic/pom.xml (hunk 2 above)
Comment (IharBakhanovich) [reply certificateshop-logic/pom.xml]: done

File: certificateshop-logic/src/main/java/com/epam/esm/configuration/LogicConfig.java (hunk 3 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/configuration/LogicConfig.java]: Done

File: certificateshop-logic/src/main/java/com/epam/esm/dao/ListToSetConverter.java (hunk 4 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/ListToSetConverter.java]: That is the point: the method converts List<List<Object>> to ResultSet.

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java (hunk 5 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java]: DEFAULT_PARAMS it not a method. It is the constant.

Question: is it reasonable to move this class to an own new package?

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java (hunk 6 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java]: Done

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java (hunk 7 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java:65]: thank you. I leave it in this form for the better code readability (for me).

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java (hunk 8 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java:79]: thank you.

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java (hunk 9 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java]: then the line has more than 120 signs. Done

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java (hunk 10 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java]: then the line has more than 120 signs. Done

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/OrderExtractor.java (hunk 11 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/OrderExtractor.java]: done

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/UserExtractor.java (hunk 12 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/UserExtractor.java]: done

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/UserExtractor.java (hunk 13 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/UserExtractor.java]: user1 is used in the line 44

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java (hunk 14 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java]: If smth. go wrong inside converter. Ii just catches this fact and throws not checked exception but unchecked.

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java (hunk 15 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java]: done

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java (hunk 16 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java]: Because i use this methods in the jdbcDao. And they are the part of the interface and should be overriden in the new realisation.
Yes, interface segregation should be done. Just now i have no idea how to implement it.

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java (hunk 17 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java]: Because i use this methods in the jdbcDao. And they are the part of the interface and should be overriden in the new realisation.
Yes, interface segregation should be done. Just now i have no idea how to implement it.

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java (hunk 18 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java]: Because I need just throw the same type of exception and i am know what can cause it (and i am sure that it won't be thrown). That is why it doesn't matter for the functionality how this exception will be named. Am i mistaking?

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java (hunk 19 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java:170]: thank you

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java (hunk 20 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java:176]: do you mean 'result' or 'listOfObjects'?

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaUserDaoImpl.java (hunk 21 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaUserDaoImpl.java:1]: Because other approaches send the queru to infinit loop. That is the point. My architecture is so. It can be upgraded in next releases potentially, but now it works the same in 'jdbc' and 'jpa' profiles. this desigion was made at the beginning of the implementation and only the one way was found to get it work properly according to the architecture of the app.

File: certificateshop-logic/src/main/java/com/epam/esm/model/converter/CertificateFromJsonConverter.java (hunk 22 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/model/converter/CertificateFromJsonConverter.java]: done

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java (hunk 16 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java]: done

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java (hunk 17 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaOrderDaoImpl.java]: done

File: certificateshop-logic/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java (hunk 23 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java]: done

File: certificateshop-logic/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java (hunk 24 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java]: done

File: certificateshop-logic/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java (hunk 25 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java]: done

File: certificateshop-logic/src/main/java/com/epam/esm/service/impl/UserServiceImpl.java (hunk 26 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/service/impl/UserServiceImpl.java]: done. I have not written tests in this module. But if it is mandatory i will.

File: certificateshop-logic/src/main/java/com/epam/esm/service/impl/UserServiceImpl.java (hunk 27 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/service/impl/UserServiceImpl.java]: done

**Review by rivulet798 — APPROVED** [review]

File: certificateshop-logic/src/main/java/com/epam/esm/dao/ListToSetConverter.java (hunk 4 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/ListToSetConverter.java]: Done

File: certificateshop-logic/src/main/java/com/epam/esm/service/impl/UserServiceImpl.java (hunk 26 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/service/impl/UserServiceImpl.java]: UserServiceTest and OrderServiceTest are written

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaUserDaoImpl.java (hunk 21 above)
Comment (IharBakhanovich) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaUserDaoImpl.java:1]: Having 'orders' by 'user' let users to make an order simultaneously with creating account (creating new user in the application). It is the one of the argumenst why this decision was made.

File: certificateshop-logic/pom.xml (hunk 1 above)
Comment (IvanYarochkin) [reply certificateshop-logic/pom.xml]: You should remove this dependency and related functionality since this is not allowed in this module.

File: certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java (hunk 5 above)
Comment (IvanYarochkin) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java]: You should move the method below the constants.

Comment (IvanYarochkin) [reply certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/ColumnNames.java]: You can keep it here

**Review by IvanYarochkin — APPROVED** [review]

**Review by IvanYarochkin — APPROVED** [review]
Remove all comments
